pkg/profile/artifact.go

```diff
@@ -140,52 +133,66 @@ func (p *Profile) getResourceIfExists(ctx context.Context, res client.Object) (b
 // MakeArtifacts creates and returns a slice of runtime.Object values, which if
 // applied to a cluster would deploy the profile as a HelmRelease.
 func (p *Profile) MakeArtifacts() ([]runtime.Object, error) {
-	objs := []runtime.Object{}
-	gitRes, helmResources, kustomizeResources, err := p.makeArtifacts()
+	objs, err := p.makeArtifacts()
 	if err != nil {
 		return nil, err
 	}
-
-	objs = append(objs, gitRes)
-	for _, helmRes := range helmResources {
-		objs = append(objs, helmRes)
-	}
-	for _, kustomizeRes := range kustomizeResources {
-		objs = append(objs, kustomizeRes)
-	}
 	return objs, nil
 }
 
-func (p *Profile) makeArtifacts() (*sourcev1.GitRepository, []*helmv2.HelmRelease, []*kustomizev1.Kustomization, error) {
-	var helmResources []*helmv2.HelmRelease
-	var kustomizeResources []*kustomizev1.Kustomization
+func (p *Profile) makeArtifacts() ([]runtime.Object, error) {
+	objs := make([]runtime.Object, 0)
+	var gitRes *sourcev1.GitRepository
 
 	for _, artifact := range p.definition.Spec.Artifacts {
+		if err := artifact.Validate(); err != nil {
+			return nil, fmt.Errorf("validation failed for artifact %s: %w", artifact.Name, err)
+		}
 		switch artifact.Kind {
-		case "HelmChart":
+		case profilesv1.HelmChartKind:
 			helmRes, err := p.makeHelmRelease(artifact)
 			if err != nil {
-				return nil, nil, nil, fmt.Errorf("failed to create HelmRelease resource: %w", err)
-
+				return nil, fmt.Errorf("failed to create HelmRelease resource: %w", err)
 			}
-			helmResources = append(helmResources, helmRes)
-		case "Kustomize":
+			objs = append(objs, helmRes)
+			if artifact.Path != "" && gitRes == nil {
+				// this resource is added at the end because it's generated once.
+				gitRes, err = p.makeGitRepository()
+				if err != nil {
+					return nil, fmt.Errorf("failed to create GitRepository resource: %w", err)
+				}
+			} else if artifact.Chart != nil {
+				helmRep, err := p.makeHelmRepository(artifact.Chart.HelmURL, artifact.Chart.HelmChart)
+				if err != nil {
+					return nil, fmt.Errorf("failed to create HelmRepository resource: %w", err)
+				}
+				objs = append(objs, helmRep)
+			}
```

Comment:
can this just be `else` rather than `else if`? Its sort of awkward, because `Validation` passed we know that its always going to work with an `if else`, but its sort of leaking that knowledge into this pkg. https://github.com/weaveworks/eksctl/issues/3566. 

Perhaps the `Artifact` struct should live here, rather than in the `profile_types` package? Just throwing random thought out.

Reply:
What do you think about just using ifs? Since the validation passed we know that only ONE should ever be there anyways.